src/NuGet.Core/NuGet.Commands/RestoreCommand/Utility/NoOpRestoreUtilities.cs

```diff
@@ -199,15 +199,31 @@ public static bool VerifyPackagesOnDisk(RestoreRequest request)
         /// <summary>
         /// Calculates the hash value, used for the no-op optimization, for the request
         /// This methods handles the deduping of tools
+        /// Handles the ignoring of RestoreSettings
         /// </summary>
         public static string GetHash(RestoreRequest request)
```

Comment:
you are modifying the object when calling GetHash which seems weird to me. Perhaps dupe the object and then call GetHash?